ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/RaceConditionIT.scala

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.suites
+
+import com.daml.ledger.api.testtool.infrastructure.Allocation._
+import com.daml.ledger.api.testtool.infrastructure.Assertions._
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+import com.daml.ledger.api.v1.transaction.{TransactionTree, TreeEvent}
+import com.daml.ledger.api.v1.value.RecordField
+import com.daml.ledger.test.semantic.RaceTests._
+import com.daml.lf.data.{Bytes, Ref}
+
+import scala.concurrent.Future
+import scala.util.Success
+
+final class RaceConditionIT extends LedgerTestSuite {
+
+  test(
+    "WWDoubleNonTransientCreate",
+    "Cannot concurrently create multiple non-transient contracts with the same key",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    val ExpectedNumberOfSuccessfulCreations = 1
+    Future
+      .traverse(1 to Attempts) { case _ =>
+        ledger.create(alice, ContractWithKey(alice)).transform(Success(_))
+      }
+      .map { results =>
+        assertLength(
+          "Successful contract creations",
+          ExpectedNumberOfSuccessfulCreations,
+          results.filter(_.isSuccess),
+        )
+        ()
+      }
+  })
+
+  test(
+    "WWDoubleArchive",
+    "Cannot archive the same contract multiple times",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    val ExpectedNumberOfSuccessfulArchivals = 1
+    for {
+      contract <- ledger.create(alice, ContractWithKey(alice))
+      _ <- Future.traverse(1 to Attempts) { _ =>
+        ledger.exercise(alice, contract.exerciseContractWithKey_Archive).transform(Success(_))
+      }
+      transactions <- ledger.transactionTrees(alice)
+    } yield {
+      import TransactionUtil._
+      assertLength(
+        "Successful contract archivals",
+        ExpectedNumberOfSuccessfulArchivals,
+        transactions.filter(isArchival),
+      )
+      ()
+    }
+  })
+
+  test(
+    "WWArchiveVsNonTransientCreate",
+    "Cannot re-create a contract if it hasn't yet been archived",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, alice)) =>
+    val Attempts = 100
+    for {
+      contract <- ledger.create(alice, ContractWithKey(alice))
+      _ <- Future.traverse(1 to Attempts) { attempt =>
+        if (attempt % 2 == 1) {
+          ledger.create(alice, ContractWithKey(alice)).transform(Success(_))
+        } else {
+          ledger.exercise(alice, contract.exerciseContractWithKey_Archive).transform(Success(_))
+        }
+      }
+      transactions <- ledger.transactionTrees(alice)
```

Comment:
Will there ever be more than 3 elements in this stream?
I believe the last 98 submissions cannot have any effect since:
- The first submission creates a contract **_contract_** with key for alice
- The second submission archives it (index 1)
- The third submission creates a **_different_** contract for the same key (index 2)
From here on, the test will not progress with contract creations since the next archive (index 3) will try to archive the original **contract** instead of the last created one. Is this the intent?

Reply:
This test is a bit different in its mechanics than the rest because here we execute `archive` 50 times and `create` 50 times.
The race condition we try to examine here is (following @mziolekda explanation):
`[create]  <wait>  [archive]-race-[create]`
In case of a bug causing the second [create] to see a partial result of [archive] command we could end up with two consecutive successful contract creations.
I've added an appropriate comment in the code.